File: algorithms/11_golang/data_structures/skip_list.go
// Package data_structures implements Skip List in Go.
package data_structures

import "math/rand"

type SkipNode struct {
	Value   int
	Forward []*SkipNode
}

type SkipList struct {
	header *SkipNode
	level  int
}

func NewSkipList() *SkipList {
	return &SkipList{
		header: &SkipNode{Forward: make([]*SkipNode, 16)},
		level:  0,
	}
}

func (sl *SkipList) Insert(val int) {
	update := make([]*SkipNode, 16)
	curr := sl.header
	for i := sl.level; i >= 0; i-- {
		for curr.Forward[i] != nil && curr.Forward[i].Value < val {
			curr = curr.Forward[i]
		}
		update[i] = curr
	}
	lvl := 0
	for rand.Float64() < 0.5 && lvl < 15 {
		lvl++
	}
	newNode := &SkipNode{Value: val, Forward: make([]*SkipNode, lvl+1)}
	for i := 0; i <= lvl; i++ {
		newNode.Forward[i] = update[i].Forward[i]
		update[i].Forward[i] = newNode
	}
}

func (sl *SkipList) Search(val int) bool {
	curr := sl.header
	for i := sl.level; i >= 0; i-- {
		for curr.Forward[i] != nil && curr.Forward[i].Value < val {
			curr = curr.Forward[i]
		}
	}
	curr = curr.Forward[0]
	return curr != nil && curr.Value == val
}

File: algorithms/04_java/data_structures/SkipList.java
package data_structures;

import java.util.Random;

/**
 * Skip List in Java (William Pugh).
 */
public class SkipList {
    private static class Node {
        int val;
        Node[] forward;
        Node(int val, int level) {
            this.val = val;
            this.forward = new Node[level + 1];
        }
    }

    private final Node head = new Node(-1, 16);
    private int level = 0;
    private final Random rand = new Random();

    public void insert(int val) {
        Node[] update = new Node[16];
        Node curr = head;
        for (int i = level; i >= 0; i--) {
            while (curr.forward[i] != null && curr.forward[i].val < val) {
                curr = curr.forward[i];
            }
            update[i] = curr;
        }
        int lvl = 0;
        while (rand.nextDouble() < 0.5 && lvl < 15) lvl++;
        Node newNode = new Node(val, lvl);
        for (int i = 0; i <= lvl; i++) {
            newNode.forward[i] = update[i].forward[i];
            update[i].forward[i] = newNode;
        }
    }

    public boolean search(int val) {
        Node curr = head;
        for (int i = level; i >= 0; i--) {
            while (curr.forward[i] != null && curr.forward[i].val < val) {
                curr = curr.forward[i];
            }
        }
        curr = curr.forward[0];
        return curr != null && curr.val == val;
    }

    public static void main(String[] args) {
        SkipList sl = new SkipList();
        sl.insert(42);
        assert sl.search(42);
        System.out.println("Java Skip List verified.");
    }
}
